crates/sui-core/src/authority/authority_store.rs

```diff
@@ -343,7 +332,59 @@ impl<S: Eq + Debug + Serialize + for<'de> Deserialize<'de>> SuiDataStore<S> {
         }
     }
 
-    pub fn get_sequenced_input_objects(
+    /// When making changes, please see if check_sequenced_input_objects() below needs
+    /// similiar changes as well.
+    pub fn get_missing_input_objects(
+        &self,
+        digest: &TransactionDigest,
+        objects: &[InputObjectKind],
+    ) -> Result<Vec<ObjectKey>, SuiError> {
+        let shared_locks_cell: OnceCell<HashMap<_, _>> = OnceCell::new();
+
+        let mut missing = Vec::new();
+        for kind in objects {
+            match kind {
+                InputObjectKind::SharedMoveObject { id, .. } => {
+                    let shared_locks = shared_locks_cell.get_or_try_init(|| {
+                        Ok::<HashMap<ObjectID, SequenceNumber>, SuiError>(
+                            self.all_shared_locks(digest)?.into_iter().collect(),
+                        )
+                    })?;
+                    match shared_locks.get(id) {
+                        Some(version) => {
+                            if !self.object_exists(id, *version)? {
+                                // When this happens, other transactions that use smaller versions of
+                                // this shared object haven't finished execution.
+                                missing.push(ObjectKey(*id, *version));
+                            }
+                        }
+                        None => {
+                            // Abort the function because the lock should have been set.
+                            return Err(SuiError::SharedObjectLockNotSetError);
+                        }
+                    };
+                }
+                InputObjectKind::MovePackage(id) => {
```

Comment:
assuming that this function only runs for certs, this should be impossible - the cert cannot have been formed if immutable inputs were missing. this may be worth a comment.

Reply:
Can users upload a move package? Is it possible a node is just behind for the move package input? Added a comment but can abort the function with an error here instead.